feat: Implement optional health server

yeldiRium opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
This MR adds a ServeOption that lets library users provide a [HealthServer](https://pkg.go.dev/google.golang.org/grpc@v1.72.1/health/grpc_health_v1#HealthServer) which will then be served with the go composition function via a common gRPC server. This allows setting up a [kubernetes readiness probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-grpc-liveness-probe) to check whether a go composition function is ready to accept gRPC connections.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

I have used this fork in our go function composition and confirmed that the passed in health server is correctly exposed via the gRPC server. I configured a kubernetes grpc liveness-probe, which was successfully called.
Since there are currently no integration or e2e tests in this project, I wasn't sure how/where to add a new test for this behavior.

I've used this minimal example for testing:
```go
package main

import (
	"github.com/crossplane/function-sdk-go"
	functionv1 "github.com/crossplane/function-sdk-go/proto/v1"
	"google.golang.org/grpc/health"
)

type Server struct {
	functionv1.UnimplementedFunctionRunnerServiceServer
}

func main() {
	server := Server{}
	healthServer := health.NewServer()

	function.Serve(
		server,
		function.Listen("tcp", ":8080"),
		function.Insecure(true),
		function.WithHealthServer(healthServer),
	)
}
```

And deployed it to kubernetes with this manifest:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: grpc-health-example
  namespace: default
spec:
  containers:
  - image: grpc-health-example:latest
    livenessProbe:
      grpc:
        port: 8080
    name: package-runtime
    ports:
    - containerPort: 8080
      protocol: TCP
```

This results in a successful liveness probe.

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

jbw976 commented:
The CI failures due to an outdated actions/cache version may have been fixed by https://github.com/crossplane/function-sdk-go/pull/199. Try to rebase your fork/branch on the latest in main and force push, then see if that helps the CI checks run again :)

jbw976 reviewed (commented):
could you also update the `How has this code been tested` with some more details? e.g. show some command outputs to show it working, how you checked, etc. 🙇‍♂️

yeldiRium commented:
Thanks for taking a look! I rebased my branch and cleaned up the history.
I'll update the description with my tests and usage examples when I get to it at the office next week.

negz commented:
I just saw this PR. It seems useful.

Do users ever need to configure the health server? I see for example https://pkg.go.dev/google.golang.org/grpc/health has Shutdown and Resume methods that I imagine the server might call to indicate health.

I ask because if users would never have to configure the health server, perhaps it should just be silently on by default. Would that work?
